libyul/backends/evm/ControlFlowGraph.h

```diff
@@ -0,0 +1,210 @@
+}
+
+/// Control flow graph consisting of ``CFG::BasicBlock``s connected by control flow.
+struct CFG
+{
+	explicit CFG() {}
+	CFG(CFG const&) = delete;
+	CFG(CFG&&) = delete;
+	CFG& operator=(CFG const&) = delete;
+	CFG& operator=(CFG&&) = delete;
+	~CFG() = default;
+
+	struct BuiltinCall
+	{
+		std::shared_ptr<DebugData const> debugData;
+		std::reference_wrapper<BuiltinFunction const> builtin;
+		std::reference_wrapper<yul::FunctionCall const> functionCall;
+		/// Number of proper arguments with a position on the stack, excluding literal arguments.
+		/// Literal arguments (like the literal string in ``datasize``) do not have a location on the stack,
+		/// but are handled internally by the builtin's code generation function.
+		size_t arguments = 0;
+	};
+	struct FunctionCall
+	{
+		std::shared_ptr<DebugData const> debugData;
+		std::reference_wrapper<Scope::Function const> function;
+		std::reference_wrapper<yul::FunctionCall const> functionCall;
+	};
+	struct Assignment
+	{
+		std::shared_ptr<DebugData const> debugData;
+		/// The variables being assigned to also occur as ``output`` in the ``Operation`` containing
+		/// the assignment, but are also stored here for convenience.
+		std::vector<VariableSlot> variables;
+	};
+
+	struct Operation
+	{
+		/// Stack slots this operation expects at the top of the stack and consumes.
+		Stack input;
+		/// Stack slots this operation leaves on the stack as output.
+		Stack output;
+		std::variant<FunctionCall, BuiltinCall, Assignment> operation;
+	};
+
+	struct FunctionInfo;
+	/// A basic control flow block containing ``Operation``s acting on the stack.
+	/// Maintains a list of entry blocks and a typed exit.
+	struct BasicBlock
+	{
+		struct MainExit {};
+		struct ConditionalJump
+		{
+			StackSlot condition;
+			BasicBlock* nonZero = nullptr;
+			BasicBlock* zero = nullptr;
+		};
+		struct Jump
+		{
+			BasicBlock* target = nullptr;
+			/// The only backwards jumps are jumps from loop post to loop condition.
+			bool backwards = false;
+		};
+		struct FunctionReturn { CFG::FunctionInfo* info = nullptr; };
+		struct Terminated {};
+		std::vector<BasicBlock*> entries;
+		std::vector<Operation> operations;
+		std::variant<MainExit, Jump, ConditionalJump, FunctionReturn, Terminated> exit = MainExit{};
+	};
+
+	struct FunctionInfo
+	{
+		std::shared_ptr<DebugData const> debugData;
+		Scope::Function const& function;
+		BasicBlock* entry = nullptr;
+		std::vector<VariableSlot> parameters;
+		std::vector<VariableSlot> returnVariables;
+	};
+
+	/// The main entry point, i.e. the start of the outermost Yul block.
+	BasicBlock* entry = nullptr;
+	/// Subgraphs for functions.
+	std::map<Scope::Function const*, FunctionInfo> functionInfo;
+	/// List of functions in order of declaration.
+	std::list<Scope::Function const*> functions;
+
+	/// Container for blocks for explicit ownership.
+	std::list<BasicBlock> blocks;
```

Comment:
Random question: is this a list instead of a vector only because you don't have to worry about pointer invalidation?

Reply:
Yep, exactly. If I made it any container that has to move stuff, the pointers in the graph could be invalidated at any time with no real way to fix.